master/buildbot/status/web/forms.py

```diff
@@ -45,55 +93,67 @@ def decodeFromURL(self, value, encoding):
     @defer.inlineCallbacks
     def content(self, request, cxt):
         status = self.getStatus(request)
-
-        args = request.args.copy()
-
-        # decode all of the args
         encoding = getRequestCharset(request)
-        for name, argl in args.iteritems():
-            if '_branch' in name:
-                args[name] = [ self.decodeFromURL(arg, encoding) for arg in argl ]
+        args = self._decodeArgs(request, encoding)
 
         #Get builder info
         builder_status = None
         if args.has_key("builder_name") and len(args["builder_name"]) == 1:
             builder_status = status.getBuilder(self.decodeFromURL(args["builder_name"][0], encoding))
-            bm = self.getBuildmaster(request)
-
-            cxt['slaves'] = [s for s in builder_status.getSlaves() if s.isConnected()]
-            cxt['slaves'] = sorted(cxt['slaves'], key=attrgetter('friendly_name'))
-
-            #Get branches
-            encoding = getRequestCharset(request)
-            branches = [ b.decode(encoding) for b in args.get("branch", []) if b ]
-            cxt['branches'] = branches
-
-            return_page = ""
-            if args.has_key("return_page"):
-                return_page = args['return_page']
-                if not isinstance(return_page, basestring):
-                    return_page = args['return_page'][0]
-
-            #Add scheduler info
-            buildForceContext(cxt, request, self.getBuildmaster(request), builder_status.getName())
-
-            #URL to force page with return param
-            url = args['builder_url'][0]
-            url_parts = list(urlparse(url))
-
-            if len(url_parts) > 0 and len(return_page) > 0:
-                return_page = "&returnpage={0}".format(return_page)
-            else:
-                return_page = "?returnpage={0}".format(return_page)
-            cxt['return_page'] = return_page
-
-            url_parts[2] += "/force"
-            url_parts[4] += return_page
-            force_url = urlunparse(url_parts)
-            cxt['force_url'] = force_url
-
-            authz = self.getAuthz(request)
-            cxt['is_admin'] = yield authz.getUserAttr(request, 'is_admin', 0)
+            buildMaster = self.getBuildmaster(request)
+
+            cxt['slaves'] = self._getSlaves(builder_status)
+            cxt['branches'] = self._getBranches(args, request)
+
+            buildForceContext(cxt, request, buildMaster, builder_status.getName())
+
+            url_parts = list(urlparse(args['builder_url'][0]))
```

Comment:
You don't have to cast result from `urlparse` to `list`, if you want to modify some part of URL you can use `_replace` method of `NamedTuple` class (https://docs.python.org/2/library/collections.html#collections.somenamedtuple._replace) and then in https://github.com/Unity-Technologies/katana/pull/290/files#diff-a3049ec7eae7a820999056008192c3a3R80 use something like this
```Python
path = urljoin(res.path, 'force')
query = res.query + return_page
return urlunparse(url_parts._replace(path=path, query=query))
```

Reply:
change it to using the url element and used replace - urljoin did not work as expected it replaces the last element instead of appending so `foo/bar/baz` becomes `foo/bar/force` instead of `foo/bar/baz/force` so I maintained the normal sting aggregation